waspls/src/Wasp/LSP/Completion.hs

```diff
@@ -0,0 +1,91 @@
+module Wasp.LSP.Completion
+  ( getCompletionsAtPosition,
+  )
+where
+
+import Control.Lens ((^.))
+import Control.Syntax.Traverse
+import Data.Maybe (maybeToList)
+import qualified Data.Text as Text
+import qualified Language.LSP.Types as LSP
+import Wasp.Backend.ConcreteSyntax (SyntaxNode)
+import qualified Wasp.Backend.ConcreteSyntax as S
+import Wasp.LSP.ServerM
+import Wasp.LSP.ServerState
+import Wasp.LSP.Syntax (findChild, isAtExprPlace, lexemeAt, positionToOffset, showNeighborhood, toOffset)
+
+-- | Get the list of completions at a (line, column) position in the source.
+getCompletionsAtPosition :: LSP.Position -> ServerM [LSP.CompletionItem]
+getCompletionsAtPosition position = do
+  src <- gets (^. sourceString)
+  maybeSyntax <- gets (^. cst)
+  case maybeSyntax of
+    -- If there is no syntax tree, make no completions
+    Nothing -> return []
+    Just syntax -> do
+      let offset = positionToOffset src position
+      -- 'location' is a traversal through the syntax tree that points to 'position'
+      let location = toOffset offset (fromSyntaxForest syntax)
+      logM $ "[getCompletionsAtPosition] neighborhood=\n" ++ showNeighborhood location
+      exprCompletions <-
+        if isAtExprPlace location
+          then do
+            logM $ "[getCompletionsAtPosition] position=" ++ show position ++ " atExpr=True"
+            getExprCompletions src syntax
+          else do
+            logM $ "[getCompletionsAtPosition] position=" ++ show position ++ " atExpr=False"
+            return []
+      let completions = exprCompletions
+      return completions
+
+-- | If the location is at an expression, find declaration names in the file
+-- and return them as autocomplete suggestions
+--
+-- TODO: include completions for enum variants (use standard type defs from waspc)
+getExprCompletions :: String -> [SyntaxNode] -> ServerM [LSP.CompletionItem]
+getExprCompletions src syntax = do
+  let declNames = findDeclNames src syntax
+  logM $ "[getExprCompletions] declnames=" ++ show declNames
+  return $
+    map
+      ( \(name, typ) ->
+          LSP.CompletionItem
+            { _label = Text.pack name,
+              _kind = Just LSP.CiVariable,
+              _tags = Nothing,
+              _detail = Just (Text.pack $ ":: " ++ typ ++ " (declaration type)"),
+              _documentation = Nothing,
+              _deprecated = Nothing,
+              _preselect = Nothing,
+              _sortText = Nothing,
+              _filterText = Nothing,
+              _insertText = Nothing,
+              _insertTextFormat = Nothing,
+              _insertTextMode = Nothing,
+              _textEdit = Nothing,
+              _additionalTextEdits = Nothing,
+              _commitCharacters = Nothing,
+              _command = Nothing,
+              _xdata = Nothing
```

Comment:
Makes sense, although because of how `lsp` overloads it's record field names, GHC can't figure out which version of `_kind` and `_detail` to use (even with `DisambiguateRecordFields`. So I had to resort to lens and ended up with:

```hs
makeBasicCompletionItem (Text.pack name)
  & (LSP.kind ?~ LSP.CiVariable)
  & (LSP.detail ?~ Text.pack (":: " ++ typ ++ " (declaration type)"))
```

which I think is still an improvement.